### PR TITLE
[FIXED JENKINS-30109] Update to bouncycastle 1.53

### PR DESCRIPTION
**Pull request by seryl**
https://issues.jenkins-ci.org/browse/JENKINS-30109

Fixes issues connecting to ECDH ssl servers /w jdk8, perhaps other fixes.

### CODE REVIEW

**Comment by recena** [comment]
@seryl Could you add an Jira issue about this?

**Comment by seryl** [comment]
https://issues.jenkins-ci.org/browse/JENKINS-30109

**Comment by recena** [comment]
@seryl Thanks so much. I'll try your proposal.

**Comment by recena** [comment]
@seryl If you are working on this ticket, it would be interesting you accept the Jira ticket.

**Comment by seryl** [comment]
@recena I'm not sure what you mean?

All of the work is done to get it finalized. Jenkins core also needed to be updated to 152, and as far as I'm aware https://github.com/jenkinsci/instance-identity-module/pull/3 does that.

I can assign the tickets to myself in the Jenkins Jira, but they are already done and I don't have merge permissions; so I'm not sure how that helps..? Let me know what further steps I need to do to pull these in.

Happy to get this finished.

**Comment by recena** [comment]
@seryl I've update [JENKINS-30109](https://issues.jenkins-ci.org/browse/JENKINS-30109) status.

**Comment by recena** [comment]
@seryl I can merge this PR but I need to review it.

**Comment by seryl** [comment]
@recena Any chance you could look at this?

36 additions.
21 deletions.

**Comment by recena** [comment]
@seryl Could you squash the commits?

File: src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgent.java [hunk 1]
```diff
@@ -83,22 +86,27 @@ public void addIdentity(String privateKey, final String passphrase, String comme
...
+                } else if (o instanceof KeyPair) {
+                    keyPair = ((KeyPair) o);
                 }
+                agent.getAgent().addIdentity(keyPair, comment);
             } finally {
                 r.close();
             }
         } catch (Exception e) {
             e.printStackTrace(listener.error(Messages.SSHAgentBuildWrapper_UnableToReadKey(e.getMessage())));
+            e.printStackTrace(listener.getLogger());
```
Comment (recena) [inline src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgent.java]: @seryl Why do you print `listener.getLogger()`?

Comment (seryl) [inline src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgent.java]: I did it to match the mirrored configuration on JNRRemoteAgent.java

https://github.com/jenkinsci/ssh-agent-plugin/blob/master/src/main/java/com/cloudbees/jenkins/plugins/sshagent/jna/JNRRemoteAgent.java#L100

If this is something you want, I assume you want it in both places.
If not, I assume you don't want it in either.

**Comment by seryl** [comment]
squashed and should be ready

**Comment by recena** [comment]
@stephenc I agree with the changes. Your feedback will be great here.

File: src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgent.java (hunk 1 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgent.java]: They should both be using the `listener.error` form

**Comment by stephenc** [comment]
Minor issue on logging. Would be great if a test could be included to prevent regressions. Otherwise looking good

**Comment by seryl** [comment]
@stephenc I'm not sure what you mean... do you want me to validate that it's printing errors when they occur?

If that's the case, I'll just remove that. I have little to no interest in getting that logging information in, I just want the bouncycastle updates. I just did it since that seemed like what the initial intention was.

File: src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgent.java (hunk 1 above)
Comment (seryl) [inline src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgent.java]: Did we not want to print the getlogger then?

Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgent.java]: `listener.error` also sends to `getLogger` but with the prefix message and allows for the two to be associated.

**Comment by stephenc** [comment]
@seryl I want you to test that ECDH keys do not blow up. A test with a test resource of an ECDH key that gets set as a credential and passed through the MinaAgent would suffice. Testing the logging information is not required.

**Comment by seryl** [comment]
I feel like there's some confusion..?

ECDH ssl is what we're adding support for here, not ecdsa keys.

We're allowing the key agreement when connecting to openssl servers to use the ecdh cipher.

This has nothing to do with openssh, so I don't understand why you are looking for a test on a new type of key?

This is to support downstream changes in the identity module.

Note: The original title was upgrade bouncycastle to 152; so that there would not be any confusion, which there seems to be.

Link to identity PR: https://github.com/jenkinsci/instance-identity-module/pull/3

When both of these are added, we'll be able to connect to apache and nginx servers which are using the openssl ciphers such as: ECDHE-ECDSA-AES256-GCM-SHA384.

The only change provided here for openssh, is to use the new bouncycastle interfaces, nothing more.

**Comment by stephenc** [comment]
Ahh ok, so you just want to bump the BC version (with the usual hassle of method signature changes)

OK... I'd prefer then if you update the PR title to reflect that as otherwise when we document the changes users will get confused just as I did

**Comment by seryl** [comment]
@stephenc Thanks for getting back to me, updated as requested

**Comment by seryl** [comment]
@stephenc https://github.com/jenkinsci/instance-identity-module/pull/3 is the associated change, this should be the last of them to allow the upgrade to 152 for vanilla jenkins

**Comment by TheSavior** [comment]
Just like https://github.com/jenkinsci/instance-identity-module/pull/3 this has been an issue for us as well and this fix works for us. Would be great to get this merged.

**Comment by seryl** [comment]
@stephenc @recena Can we please give this some attention? It's been over 2 months now.

**Comment by stephenc** [comment]
Still waiting for the requested changes to be made

**Comment by seryl** [comment]
I'm not sure I follow. What requested changes?

**Comment by stephenc** [comment]
You removed the correct way of logging the stack trace in one of the lines you changed. I requested you to revert that change in the PR comment on that line

**Comment by seryl** [comment]
@stephenc removed as requested.

**Pull request merged**